code/game/objects/structures/shower.dm

```diff
@@ -114,26 +117,26 @@
 
 /obj/machinery/shower/Crossed(atom/movable/AM)
 	..()
-	if(on)
+	if(on && reagents.total_volume)
 		wash_atom(AM)
 
-/obj/machinery/shower/proc/wash_atom(atom/A)
-	A.wash(CLEAN_RAD | CLEAN_TYPE_WEAK) // Clean radiation non-instantly
-	A.wash(CLEAN_WASH)
-	SEND_SIGNAL(A, COMSIG_ADD_MOOD_EVENT, "shower", /datum/mood_event/nice_shower)
-	reagents.expose(A, TOUCH)
-	if(isliving(A))
-		check_heat(A)
+/obj/machinery/shower/proc/wash_atom(atom/target)
+	target.wash(CLEAN_RAD | CLEAN_TYPE_WEAK) // Clean radiation non-instantly
+	target.wash(CLEAN_WASH)
+	SEND_SIGNAL(target, COMSIG_ADD_MOOD_EVENT, "shower", /datum/mood_event/nice_shower)
+	reagents.expose(target, (TOUCH), SHOWER_EXPOSURE_MULTIPLIER * SHOWER_SPRAY_VOLUME / max(reagents.total_volume, SHOWER_SPRAY_VOLUME))
+	if(isliving(target))
+		check_heat(target)
 
 /obj/machinery/shower/process()
-	if(on && reagents.total_volume >= 5)
-		reagents.remove_any(5)
+	if(on && reagents.total_volume)
 		wash_atom(loc)
 		for(var/am in loc)
 			var/atom/movable/movable_content = am
-			reagents.expose(movable_content, TOUCH, SHOWER_REACTION_MULTIPLIER) //There's not many reagents leaving the sink at once! This should make for a 10 unit reaction
 			if(!ismopable(movable_content)) // Mopables will be cleaned anyways by the turf wash above
-				wash_atom(movable_content)
+				wash_atom(movable_content)	// Reagent exposure is handled in wash_atom
+
+		reagents.remove_any(SHOWER_SPRAY_VOLUME)
```

Comment:
Shouldn't this also be in wash_atom due to the Crossed() call?

Reply:
Not exactly. That would make the number of objects the shower is washing change the amount of reagents the shower drains. That's not exactly how shower works and I don't feel like dealing with the race conditions would be worth it.